packages/storefront-lwc/src/modules/commerce/basket/__tests__/basket.test.js

```diff
@@ -0,0 +1,276 @@
+import Basket from 'commerce/basket';
+import { mockQuery, reset } from '@lwce/apollo-client';
+import { createElement } from 'lwc';
+
+import { registerWireService } from '@lwc/wire-service';
+import { register } from 'lwc';
+registerWireService(register);
+
+function tick(time = null) {
+    return new Promise(resolve => setTimeout(resolve, time));
+}
+
+describe('<commerce-basket>', function() {
+    afterEach(() => {
+        // The jsdom instance is shared across test cases in a single file so reset the DOM
+        while (document.body.firstChild) {
+            document.body.removeChild(document.body.firstChild);
+        }
+
+        reset();
+    });
+
+    it('should render basket loading', async function() {
+        const element = createElement('commerce-basket', {
+            is: Basket,
+        });
+
+        document.body.appendChild(element);
+
+        await tick();
+
+        expect(element.shadowRoot.querySelector('.loading')).toMatchSnapshot();
+    });
+
+    it('should render an empty basket', async function() {
+        mockQuery(emptyBasketResponse);
+
+        const element = createElement('commerce-basket', {
+            is: Basket,
+        });
+
+        document.body.appendChild(element);
+
+        await tick();
+
+        expect(
+            element.shadowRoot.querySelector('.text-empty-basket'),
+        ).toMatchSnapshot();
+    });
+
+    it('should render a basket with an item', async function() {
+        mockQuery(basketWithItem);
+
+        const element = createElement('commerce-basket', {
+            is: Basket,
+        });
+
+        document.body.appendChild(element);
+
+        await tick();
+
+        expect(
+            element.shadowRoot.querySelector('.header-title .number-of-items'),
+        ).toMatchSnapshot();
+    });
+
+    it('should update basket details', async function() {
+        mockQuery(basketWithItem);
+
+        const element = createElement('commerce-basket', {
+            is: Basket,
+        });
+
+        document.body.appendChild(element);
+
+        await tick();
+
+        element.shadowRoot
+            .querySelector('commerce-shippingmethods')
+            .dispatchEvent(
+                new CustomEvent('updateshippingmethod', {
+                    detail: {
+                        updatedBasket: {
+                            ...basketWithItem.data.getBasket,
+                            totalProductsQuantity: 2,
+                        },
+                    },
+                }),
+            );
+
+        await tick();
+
+        expect(
+            element.shadowRoot.querySelector('.header-title .number-of-items'),
+        ).toMatchSnapshot();
+    });
+});
+
+const emptyBasketResponse = {
```

Comment:
This is literally a copy and pasted response from the server

Reply:
Should we move these to the api level so that other components that need basket response can use these?

Reply:
I pulled it out and put it into a testData file. We need to be careful though, when multiple tests reuse the same data, it is really important that tests don't mutate the data, otherwise you can get stability issues between test runs where tests don't expect someone else is modifying things.